cmd/csi-proxy/main.go

```diff
@@ -68,3 +84,56 @@ func apiGroups() ([]srvtypes.APIGroup, error) {
 		smbsrv,
 	}, nil
 }
+
+// configure as a Windows service managed by Windows SCM
+// code borrowed from
+// https://github.com/kubernetes/kubernetes/blob/323f34858de18b862d43c40b2cced65ad8e24052/pkg/windows/service/service.go
+func initService() error {
+	h := &handler{
+		tosvc:   make(chan bool),
+		fromsvc: make(chan error),
+	}
+
+	service = h
+	var err error
+	go func() {
+		err = svc.Run("csiproxy", h)
+		h.fromsvc <- err
+	}()
+
+	// Wait for the first signal from the service handler.
+	err = <-h.fromsvc
+	if err != nil {
+		return err
+	}
+	klog.Infof("Running as a Windows service.")
+	return nil
+}
+
+func (h *handler) Execute(_ []string, r <-chan svc.ChangeRequest, s chan<- svc.Status) (bool, uint32) {
+	s <- svc.Status{State: svc.StartPending, Accepts: 0}
+	// Unblock initService()
+	h.fromsvc <- nil
+
+	s <- svc.Status{State: svc.Running, Accepts: svc.AcceptStop | svc.AcceptShutdown | svc.Accepted(windows.SERVICE_ACCEPT_PARAMCHANGE)}
+	klog.Infof("Windows Service initialized through SCM")
+Loop:
+	for {
+		select {
+		case <-h.tosvc:
+			break Loop
+		case c := <-r:
+			switch c.Cmd {
+			case svc.Interrogate:
+				s <- c.CurrentStatus
+			case svc.Stop, svc.Shutdown:
+				// todo: need to add a ctx to servers
```

Comment:
This will need a fair bit of re-architecture of server.go to pass down [1] a cancellable `context` and `waitgroup` to every server across every API version and [2] spin up each server under the context of an `errgroup` to handle errors. Will address this in a later PR as this affects only the service shutdown path for now.